x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/jinaai/JinaAIResponseHandler.java

```diff
@@ -0,0 +1,98 @@
+    static final String INPUTS_ARRAY_ERROR_MESSAGE = "Received an inputs array too large response";
+    static final String PAYMENT_ERROR_MESSAGE = "Payment required";
+
+    public JinaAIResponseHandler(String requestType, ResponseParser parseFunction) {
+        super(requestType, parseFunction, JinaAIErrorResponseEntity::fromResponse);
+    }
+
+    @Override
+    public void validateResponse(ThrottlerManager throttlerManager, Logger logger, Request request, HttpResult result)
+        throws RetryException {
+        checkForFailureStatusCode(request, result);
+        checkForEmptyBody(throttlerManager, logger, request, result);
+    }
+
+    @Override
+    public boolean canHandleStreamingResponses() {
+        return false;
+    }
+
+    @Override
+    public InferenceServiceResults parseResult(Request request, Flow.Publisher<HttpResult> flow) {
+        // TODO (JoanFM): Check this logic
+        return new StreamingChatCompletionResults(null);
+    }
+
+    /**
+     * Validates the status code throws an RetryException if not in the range [200, 300).
+     *
+     * @param request The http request
+     * @param result  The http response and body
+     * @throws RetryException Throws if status code is {@code >= 300 or < 200 }
+     */
+    @Override
+    protected void checkForFailureStatusCode(Request request, HttpResult result) throws RetryException {
+        if (result.isSuccessfulResponse()) {
+            return;
+        }
+
+        // handle error codes
+        int statusCode = result.response().getStatusLine().getStatusCode();
+        if (statusCode == 500) {
+            throw new RetryException(true, buildError(SERVER_ERROR, request, result));
+        } else if (statusCode > 500) {
+            throw new RetryException(false, buildError(SERVER_ERROR, request, result));
+        } else if (statusCode == 429) {
+            throw new RetryException(true, buildError(RATE_LIMIT, request, result));
+        } else if (statusCode == 400 && isTextsArrayTooLarge(result, statusCode)) {
+            throw new RetryException(false, buildError(INPUTS_ARRAY_ERROR_MESSAGE, request, result));
+        } else if (statusCode == 401) {
+            throw new RetryException(false, buildError(AUTHENTICATION, request, result));
+        } else if (statusCode == 402) {
+            throw new RetryException(false, buildError(PAYMENT_ERROR_MESSAGE, request, result));
+        } else {
+            throw new RetryException(false, buildError(UNSUCCESSFUL, request, result));
+        }
+    }
+
+    private static boolean isTextsArrayTooLarge(HttpResult result, int statusCode) {
+
+        if (statusCode == 400) {
+            var errorEntity = JinaAIErrorResponseEntity.fromResponse(result);
+
+            return errorEntity != null && errorEntity.getErrorMessage().contains(INPUTS_ARRAY_TOO_LARGE_MESSAGE_MATCHER);
+        }
+
+        return false;
+    }
```

Comment:
From a JinaAI perspective it probably makes sense to handle these specific errors, but from Elastic's perspective it probably makes more sense to stay generic and do not check concrete error messages as these could change at any time without us noticing. I guess your API returns anyway a useful error? (assuming it as you're checking for one :) )


I suggest to remove the error message specific response handling. WDYT @jonathan-buttner ?

Reply:
I saw it from `Cohere`. But I agree with the thing that we should not depend on this